scripts/apps/authoring-react/macros/macros.tsx

```diff
@@ -0,0 +1,213 @@
+
+    return sdApi.article.patch(
+        currentArticle,
+        patchCopy,
+        {patchDirectlyAndOverwriteAuthoringValues: true},
+    ).then(() => {
+        dispatchInternalEvent('dangerouslyForceReloadAuthoring', null);
+    });
+}
+
+class MacrosWidget extends React.PureComponent<IProps, IState> {
+    constructor(props: IProps) {
+        super(props);
+
+        this.state = {
+            macros: null,
+            article: {
+                saved: null,
+                autosaved: null,
+            },
+            displayGrouped: false,
+        };
+    }
+
+    componentDidMount(): void {
+        Promise.all([
+            this.getAllMacros(),
+            this.props.authoringStorage.getEntity(this.props.article._id),
+        ]).then(([macros, article]) => {
+            const groupedMacros = groupBy(macros._items.filter((x) => x.group != null), nameof<IMacro>('group'));
+
+            this.setState({
+                macros: macros._items,
+                article: article,
+                displayGrouped: Object.keys(groupedMacros).length > 0 ? true : null,
+            });
+        });
+    }
+
+    getAllMacros(page = 1): Promise<IRestApiResponse<IMacro>> {
+        return httpRequestJsonLocal<IRestApiResponse<IMacro>>({
+            method: 'GET',
+            path: '/macros',
+            urlParams: {
+                backend: false,
+                desk: sdApi.desks.getCurrentDeskId(),
+                max_results: 200,
+                page: page,
+            },
+        });
+    }
+
+    runMacro(macro: IMacro) {
+        return httpRequestJsonLocal({
+            method: 'POST',
+            path: '/macros',
+            payload: {
+                macro: macro.name,
+                item: this.state.article.saved,
+            },
+        }).then((res: IMacro) => {
+            const patch = generatePatch(this.props.article, res.item);
+            const isEmpty = Object.keys(patch).length < 1;
+
+            if (!isEmpty) {
+                overwriteArticle(this.props.article, patch, this.props.contentProfile.content);
+            }
+        });
+    }
+
+    render() {
+        if (this.state.macros == null) {
+            return null;
+        }
+
+        const groupedOrdered: Dictionary<string, Array<IMacro>> = {};
+
+        if (this.state.displayGrouped != null) {
+            groupedOrdered['Quick List'] = this.state.macros.filter((m) => m.order != null);
+            const groupedMacros = groupBy(this.state.macros.filter((m) => m.group != null), nameof<IMacro>('group'));
+
+            Object.keys(groupedMacros).sort().forEach((key) => {
+                groupedOrdered[key] = groupedMacros[key].sort((a, b) => a.label.localeCompare(b.label));
+            });
+            groupedOrdered['Miscallaneous'] = this.state.macros.filter((x) => x.group == null);
+        }
+
+        const RunMacroButton = (macro: IMacro) => {
+            return (
+                <div key={macro.name} style={{paddingTop: 4}}>
```

Comment:
It's not a good practice to use key outside of iteration. Better convert this to an actual component and use it like that:
 `<RunMacroButton key={macro.name} macro={macro} />`